File: 560-subarray-sum-equals-k/560-subarray-sum-equals-k.js
/**
 * @param {number[]} nums
 * @param {number} k
 * @return {number}
 */
var subarraySum = function(nums, k) {
  const cache = {}
  cache[0] = 1
  let sum = 0
  let count = 0
  for (const n of nums) {
    sum += n
    if (cache[sum-k]) {
      count += cache[sum-k]
    }
    if (cache[sum]) {
      cache[sum]++
    } else {
      cache[sum]  = 1
    }
  }
  return count
}

File: 560-subarray-sum-equals-k/560-subarray-sum-equals-k.go
func subarraySum(nums []int, k int) int {
  res := 0
  cache := make(map[int]int, len(nums))
  cache[0] = 1
  sum := 0
  for i := 0; i < len(nums); i++ {
    sum += nums[i]
    res += cache[sum - k]
    cache[sum]++
  }
  return res
}